[C#] New Concept Exercise: arrays (#1257)
exercism/v3: languages/csharp/exercises/concept/arrays/.docs/after.md

> Data structures that can hold zero or more elements are known as _collections_. An **array** is a collection that has a fixed size and whose elements must all be of the same type. Elements can be assigned to an array or retrieved from it using an index, with the first element's index being zero:

Author: ```suggestion
Data structures that can hold zero or more elements are known as _collections_. An **array** is a collection that has a fixed size and whose elements must all be of the same type. Elements can be assigned to an array or retrieved from it using an index. C# arrays are zero-based, meaning that the first element's index is always zero:
```

> ```csharp

Reviewer: hmmm... just noticed that this is the `after` document, not the introduction one... 
why do we need re-introduce all concepts (including detailed  a syntax description) for an exercise the student just finished?

Reviewer: It's also weird that this document contains more details than the `introduction` one...

Author: @maurelio1234 Well that is precisely what the `after.md` document should do. This comment to me seems to imply that you're not yet familiar with what the contents of the documents should be. Therefore, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

Reviewer: Yes, I saw these documents before posting (not the video though) and a couple of other exercises... but it is still weird to me to give more details after then before.

I see the point of summarizing what was presented in the exercise (as it is said here: https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#docsaftermd) but not of presenting the same thing again with more details... it seems too late for that... 

But again :) thanks for the pointers I will see the video.

Author: > but it is still weird to me to give more details after then before.

Well, we disagree :) The thing we want is for a student to learn a concept by themselves. This means providing them with the minimal amount of information necessary to solve the exercise. Once the student has demonstrated the mastery of the concepts by completing the exercise, the student is then ready to dig a bit deeper into the concept, for example by showing why things work a specific way or by linking to more in-depth tutorials.

Hopefully the video will help explain things.

Reviewer: > Once the student has demonstrated the mastery of the concepts by completing the exercise, the student is then ready to dig a bit deeper into the concept, for example by showing why things work a specific way or by linking to more in-depth tutorials.

Oh yes, I see ;) It makes sense now ;)

// Declare array with explicit size (size is 2)
int[] twoInts = new int[2];

// Assign first and second element by index
twoInts[0] = 7;
twoInts[1] = 8;

// Retrieve the second element by index
twoInts[1] == 8; // => true

// Check the length of the array
twoInts.Length == 2; // => true
```

Arrays can also be defined using a shortcut notation that allows you to both create the array and set its value. As the compiler can now tell how many elements the array will have, the length can be omitted:

```csharp
// Assign and initialize (size is 3)
int[] threeIntsV1 = new int[] { 4, 9, 7 };

// Shorter notation (size is 3)
int[] threeIntsV2 = new[] { 4, 9, 7 };

// Shortest notation (size is 3)
int[] threeIntsV3 = { 4, 9, 7 };
```

> Arrays can be manipulated by either calling an array's [methods][array-methods] or [properties][array-properties], or by using the methods defined in the [`Array` class][array-class].

Reviewer: ```suggestion
Arrays can be manipulated by using the methods defined in the [`Array` class][array-class].
```
the linked methods and properties are the ones on the array class...

Author: Eh, yes, and that is also precisely what I _want_ to link to. See also https://github.com/exercism/v3/pull/1257#discussion_r412244208

Reviewer: As I said on the other comment ... so by saying  `or by using the methods defined in the [`Array` class][array-class].`  do you mean using the static methods defined at the class level?

English is not my first language, but it looks like you just saying the same thing twice

- `calling an array's [methods][array-methods] or [properties][array-properties],`
-  `using the methods defined in the [`Array` class][array-class].`

There's some non written information here, and I am not sure students will understand what it is (I don't).

Author: Ah right, I didn't get it, but now I do! I've updated the text to make the distinction more clear.

An array is also a _collection_, which means that you can iterate over all its values using a [`foreach` loop][foreach-statement]:

```csharp

> char[] vowels = new [] { 'a', 'e', 'i', 'o', 'u' }; // Size is 5

Reviewer: Why repeat `Size is X` in many places? and in different formats? (up it is `(size is X)`
To me it should be used when it's relevant... and omitted otherwise.

Author: Agreed. I've omitted it here and shortened the other example.

foreach (char vowel in vowels)
{
    // Output the vowel
    System.Console.Write(vowel);
}

// => aeiou
```

One could use a [`for` loop][for-statement] to iterate over an array:

```csharp
char[] vowels = new [] { 'a', 'e', 'i', 'o', 'u' };

for (int i = 0; i < vowels.Length; i++)
{
    // Output the vowel
    System.Console.Write(vowels[i]);
}

// => aeiou
```

However, generally a `foreach` loop is preferrable over a `for` loop for the following reasons:

- A `foreach` loop is guaranteed to iterate over _all_ values. With a `for` loop, it is easy to miss elements, for example due to an off-by-one error.
- A `foreach` loop is more _declarative_, your code is communicating _what_ you want it to do, instead of a `for` loop that communicates _how_ you want to do it.

> - A `foreach` loop works on all collection types, includding those that don't support using an indexer to access elements.

Reviewer 2: ```suggestion
- A `foreach` loop works on all collection types, including those that don't support using an indexer to access elements.
```

To guarantee that a `foreach` loop will iterate over _all_ values, the compiler will not allow updating of a collection within a `foreach` loop:

```csharp
char[] vowels = new [] { 'a', 'e', 'i', 'o', 'u' };

foreach (char vowel in vowels)
{
    // This would result in a compiler error
    // vowel = 'Y';
}
```

> If you want to stop iterating when using a `foreach` loop, the [`break` keyword][break-keyword] can be used:

Reviewer: hmmm... it's the same as for all kinds of loop right?
why presenting `break` and not `continue` for example?

Author: Good point. I don't want to overload the student with information, so I'll just remove the `break` bit and explain that in a different exercise.
